rules/key-principles-of-rest-api-design/rule.md

```diff
@@ -0,0 +1,104 @@
+    url: https://ssw.com.au/people/jeoffrey-fischer
+related:
+- choose-the-right-api-tech
+- do-you-know-the-common-design-principles-part-1
+redirects: []
+
+---
+
+REST APIs are everywhere in our daily lives—from social media updates and online shopping to weather forecasts and GPS navigation. These interfaces allow different applications to connect seamlessly, providing the backbone for countless modern conveniences.
+
+Building a high-quality API means adhering to best practices that enhance clarity, reliability, and security, creating a more consistent experience for developers and users.
+
+<!--endintro-->
+
+`youtube: https://youtu.be/_gQaygjm_hg`
+**Video: Good APIs Vs Bad APIs: 7 Tips for API Design (6 min)**
+
+## Tip \#1 - Use clear naming  
+
+Choose descriptive, intuitive names for endpoints and parameters to improve readability. Adhere to REST conventions, using nouns and verbs logically (e.g., `/api/users` for accessing user data, `/api/users/{id}` for specific user information). Endpoints should also use plurals. i.e `/api/users` instead of `/api/user`. This provides a much more consistent API structure when querying both collections and single entities.
+
+Clear naming simplifies usage and reduces misunderstandings.  
+
+## Tip \#2 - Ensure reliability through idempotency  
+
+To avoid unintended actions from repeated requests, design POST, PUT, and DELETE operations as idempotent, where repeating an action yields the same result as performing it once.
+
+For example, if a DELETE request removes a record, re-sending it should not throw errors if the record is already deleted.
+
+This prevents accidental duplicate data processing.  
+
+## Tip \#3 - Implement versioning  
+
+Introduce versioning from the start (e.g., `/v1/resource`) to maintain backward compatibility when updating the API.
+
+Versioning helps users manage changes without breaking existing implementations, allowing them to adopt new features gradually.  
+
+## Tip \#4 - Add pagination for responses  
+
+For endpoints that return lists, apply pagination to prevent overwhelming the client with too much data.
+
+When paging parameters are omitted from the request, the API should apply some sensible defaults (e.g. page 1, 50 records).
+
+Use query parameters like ?page= and ?limit= to specify page numbers and size, offering a more manageable data experience while improving performance.  
+
+## Tip \#5 - Use clear query strings for sorting  
+
+When supporting sorting, apply clear query strings (e.g., `?sortBy=name&order=asc`). Query params should generally be optional making the API easier to consume.
+
+Consistent sorting parameters allow developers to retrieve and organize data efficiently and minimize confusion in handling API responses.  
+
+## Tip \#6 - Security should not be an afterthought
+
+Security is paramount in API design. Pass sensitive information, like access tokens, securely via headers (e.g., `Authorization: Bearer {token}`).  
+
```

Comment:
I don't like this sentence. It implies that HTTP headers are more "secure" than alternatives (body? query params?).

Headers are just as insecure as anything else. I think a better phrase would be something like:

Security is paramount when building REST APIs. Most REST APIs are hosted online, and you don't want to rely on "security via obscurity". Ensure you spend time hardening your surface area.
Common sense approaches include protecting your endpoints via short-lived access tokens (even for seemingly benign functionality), as well as your typical security headers such as:

Reply:
Done - I've updated the comment